Support repeated elements in XML configuration, this addresses #745 and #815

amoerie opened the pull request:
- Maintain a stack of elements that are encountered during traversal of the XML file
- Detect sibling elements and automatically append indexes to the generated configuration keys, exactly like how the JSON configuration providers does for JSON arrays
- Add a battery of unit tests to verify no existing features were broken in the process

I know this is a considerable change in the existing code, but all existing tests were not modified and are still green. The "Name" feature is still completely supported and as such, this is not a breaking change.

However, adding a "Name" is no longer required for repeated XML elements, the keys will be automatically populated with indexes instead. This is exactly how the JSON provider handles JSON arrays.

Feedback is most welcome!

Thank you for your consideration.

Commit:
Add support for repeated XML elements without a Name attribute
- Maintain a stack of elements that are encountered during traversal of the XML file
- Detect sibling elements and automatically append indexes to the generated configuration keys, exactly like how the JSON configuration providers does for JSON arrays
- Add a battery of unit tests to verify no existing features were broken in the process

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/Configuration?pullRequest=816) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/Configuration?pullRequest=816) before we can accept your contribution.<br/><br/>:x: amoerie [sign now](https://cla.dotnetfoundation.org/aspnet/Configuration?pullRequest=816)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/Configuration?pullRequest=816) it.</sub>

amoerie commented on src/Config.Xml/XmlConfigurationProvider.cs:
```diff
@@ -180,41 +180,30 @@ private void ProcessAttributes(XmlReader reader, Stack<Element> elementPath, ILi
-            data[key] = reader.Value;
-            prefixStack.Pop();
+            return name;
         }
```
It also turned out that this method wasn't used anymore, so it's been removed

Commit:
Read Name attribute case insensitively
- Add a unit test that covers this scenario (previously only functional tests failed)
- Eagerly load the name attribute when an element is encountered, because it is needed for sibling detection.

Commit:
Remove redundant null check in while loop when getting the 'Name' attribute

ajcvickers commented:
@amoerie After some discussion, we have decided that this would be a good PR to take, since it improves the XML experience with little risk of breaking real-world configs that already exist.

We have recently consolidated the configuration code into the https://github.com/aspnet/Extensions repo. Unfortunately, it is not trivial to move PRs over to another repo, so will not be moving this PR over. However, it should not be too difficult to recreate in the new location if you wish to continue pursuing it.
